FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,91 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				let fetchedFeedRequest: NSFetchRequest<CoreDataFeed> = CoreDataFeed.fetchRequest()
+				let result = try context.fetch(fetchedFeedRequest)
+
+				guard let coreDataFeed = result.first,
+				      let timestamp = coreDataFeed.timestamp,
+				      let coreDataFeedArray = (coreDataFeed.coreDataFeedImages?.array as? [CoreDataFeedImage]) else {
+					completion(.empty)
+					return
+				}
+
+				completion(.found(feed: coreDataFeedArray.toSortedLocal(), timestamp: timestamp))
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.performAndWait { [weak self] in
+
+			do {
+				try self?.deleteCacheIfExists(on: context)
+			} catch {
+				context.rollback()
+				completion(error)
+				return
+			}
+
+			let coreDataFeedEntity = CoreDataFeed(context: context)
+			coreDataFeedEntity.timestamp = timestamp
+
+			let entities: [CoreDataFeedImage] = feed.map { item in
+				let entity = CoreDataFeedImage(context: context)
+				entity.url = item.url
+				entity.id = item.id
+				entity.location = item.location
+				entity.imageDescription = item.description
+				entity.coreDataFeed = coreDataFeedEntity
+				return entity
+			}
+
+			coreDataFeedEntity.coreDataFeedImages = NSOrderedSet(array: entities)
+
+			do {
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.performAndWait { [weak self] in
```

Comment:
There's no need to weakify `self` in this context. Remove `[weak self]`.

Moreover, `performAndWait` will block clients. Use `perform` instead.

```suggestion
		context.perform {
```

Reply:
fixed